src/foam/nanos/ruler/Rule.js

```diff
@@ -180,7 +192,9 @@
       javaCode: `
         getAsyncAction().applyAction(x, obj, oldObj, ruler);
         if ( ! getAfter()
-          && Operations.REMOVE != getOperation()
+          && Operations.CREATE == getOperation()
+          || Operations.UPDATE == getOperation()
+          || Operations.CREATE_OR_UPDATE == getOperation()
         ) {
           ruler.getDelegate().put_(x, obj);
```

Comment:
this is an asynchronous, applyAction should return a promise so you can perform the getDelegate.put on it's completion.  Or does applyAction already do this when it completes?

Reply:
This block will be run inside a thread pool on the server side. https://github.com/lchanmann/foam2/blob/RulerEnhancementPOC_part3/src/foam/nanos/ruler/RuleEngine.java#L102.